fs/parallel_unpacker.go

```diff
@@ -0,0 +1,143 @@
+   limitations under the License.
+*/
+
+package fs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/errgroup"
+)
+
+const defaultRetryPeriod = 500 * time.Millisecond
+
+type parallelLayerUnpacker struct {
+	*layerUnpacker
+	controller            LayerUnpackResourceController
+	discardUnpackedLayers bool
+}
+
+func NewParallelLayerUnpacker(fetcher Fetcher, archive Archive, controller LayerUnpackResourceController, discardUnpackedLayers bool) *parallelLayerUnpacker {
+	return &parallelLayerUnpacker{
+		layerUnpacker: &layerUnpacker{
+			fetcher: fetcher,
+			archive: archive,
+		},
+		controller:            controller,
+		discardUnpackedLayers: discardUnpackedLayers,
+	}
+}
+
+func (lu *parallelLayerUnpacker) Unpack(ctx context.Context, desc ocispec.Descriptor, mountpoint string, mounts []mount.Mount) error {
+	rc, local, err := lu.fetcher.Fetch(ctx, desc)
+	if err != nil {
+		return fmt.Errorf("cannot fetch layer: %w", err)
+	}
+	defer rc.Close()
+
+	release, err := lu.controller.AcquireUnpackLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer release()
+
+	var errGroup errgroup.Group
+	if !local && !lu.discardUnpackedLayers {
+		var fetchLocalContentStoreReader bool
+
+		ingestReader, err := lu.controller.GetUnpackIngestReader()
+		if err != nil && errors.Is(err, ErrLayerIngestDoesNotExist) {
+			// Compressed layer is not on disk; so use the remote reader to download and store it in the content store.
+			// Close has already been deferred above.
+			fetchLocalContentStoreReader = true
+			ingestReader = rc
+		} else if err != nil {
+			return fmt.Errorf("cannot get layer ingest: %w", err)
+		} else {
+			defer ingestReader.Close()
+		}
+
+		errGroup.Go(func() error {
+			ticker := time.NewTicker(defaultRetryPeriod)
+			defer ticker.Stop()
+
+			err := lu.fetcher.Store(ctx, desc, ingestReader)
+			// Loop every half second until we can get the lock
+			for errdefs.IsUnavailable(err) {
+				select {
+				case <-ctx.Done():
+					err = ctx.Err()
+					return err
+				case <-ticker.C:
+					err = lu.fetcher.Store(ctx, desc, ingestReader)
+				}
+			}
+			if err != nil {
+				return fmt.Errorf("cannot store layer at %s: %w", mountpoint, err)
+			}
+			log.G(ctx).WithField("digest", desc.Digest).WithField("mountpoint", mountpoint).Debug("Layer stored")
+			return nil
+		})
+
+		if fetchLocalContentStoreReader {
+			// Wait for content to be stored in the content store before fetching it.
+			// This is to ensure that the content is available when we fetch it.
+			if err := errGroup.Wait(); err != nil {
+				return err
+			}
+
+			rc, _, err = lu.fetcher.Fetch(ctx, desc)
+			if err != nil {
+				return fmt.Errorf("cannot fetch layer: %w", err)
+			}
+			defer rc.Close()
+		}
```

Comment:
Not sure I understand this piece. Isn't it effectively making the content store writing and layer unpacking sequential?

Reply:
Artifact from earlier versions of the code. We can remove a lot of this logic and just store them in parallel